test/modules/TestResultCoordinator/Reports/DirectMethod/LongHaul/DirectMethodLongHaulReportGenerator.cs

```diff
@@ -0,0 +1,113 @@
+            string trackingId,
+            string senderSource,
+            ITestResultCollection<TestOperationResult> senderTestResults,
+            Option<string> receiverSource,
+            Option<ITestResultCollection<TestOperationResult>> receiverTestResults,
+            string resultType)
+        {
+            if (receiverSource.HasValue ^ receiverTestResults.HasValue)
+            {
+                throw new ArgumentException("Provide both receiverSource and receiverTestResults or neither.");
+            }
+
+            this.TestDescription = Preconditions.CheckNonWhiteSpace(testDescription, nameof(testDescription));
+            this.trackingId = Preconditions.CheckNonWhiteSpace(trackingId, nameof(trackingId));
+            this.SenderSource = Preconditions.CheckNonWhiteSpace(senderSource, nameof(senderSource));
+            this.SenderTestResults = Preconditions.CheckNotNull(senderTestResults, nameof(senderTestResults));
+            this.ReceiverSource = receiverSource;
+            this.ReceiverTestResults = receiverTestResults;
+            this.ResultType = Preconditions.CheckNonWhiteSpace(resultType, nameof(resultType));
+        }
+
+        internal Option<string> ReceiverSource { get; }
+
+        internal Option<ITestResultCollection<TestOperationResult>> ReceiverTestResults { get; }
+
+        internal string SenderSource { get; }
+
+        internal ITestResultCollection<TestOperationResult> SenderTestResults { get; }
+
+        internal string ResultType { get; }
+
+        internal string TestDescription { get; }
+
+        internal ITestResultComparer<TestOperationResult> TestResultComparer { get; }
+
+        public async Task<ITestResultReport> CreateReportAsync()
+        {
+            ulong senderSuccesses = 0;
+            Option<ulong> receiverSuccesses = Option.None<ulong>();
+            ulong statusCodeZero = 0;
+            ulong unknown = 0;
+            while (await this.SenderTestResults.MoveNextAsync())
+            {
+                this.ValidateDataSource(this.SenderTestResults.Current, this.SenderSource);
+                DirectMethodTestResult dmSenderTestResult = JsonConvert.DeserializeObject<DirectMethodTestResult>(this.SenderTestResults.Current.Result);
+                switch ((int)dmSenderTestResult.HttpStatusCode)
+                {
+                    case 0:
+                        statusCodeZero++;
+                        break;
+                    case 200:
+                        senderSuccesses++;
+                        break;
+                    default:
+                        unknown++;
+                        break;
+                }
+            }
+
+            await this.ReceiverTestResults.ForEachAsync(async r =>
+            {
+                ulong receiverResults = 0;
+                while (await r.MoveNextAsync())
+                {
+                    // ReceiverSource will always be there if ReceiverTestResults is so it's safe to put OrDefault
```

Comment:
What default will this give if we are wrong? It might be better to try to unwrap and throw an exception if none.